AnalyticsRallyHere/Source/AnalyticsRallyHere/Private/RH_AnalyticsProviderEventCache.cpp

```diff
@@ -221,14 +236,139 @@ FRH_AnalyticsProviderEventCache::FRH_AnalyticsProviderEventCache(int32 InMaximum
 	RHEventCacheStatic::InitializePayloadBuffer(CachedEventUTF8Stream, PreallocatedPayloadSize);
 }
 
-// We start with {"eventList":[]}
-// We End with {"eventList":[{"eventName":"<NAME>","eventTimestamp":"<timestamp>","eventUUID":"<guid>",<DefaultAttrs>,"eventParams": {<Attrs>}}]}
+void FRH_AnalyticsProviderEventCache::SetUserId(TOptional<FString> InUserId)
+{
+	FScopeLock ScopedLock(&CachedEventsCS);
+	UserId = MoveTemp(InUserId);
+}
+
+TOptional<FString> FRH_AnalyticsProviderEventCache::GetUserId() const
+{
+	FScopeLock ScopedLock(&CachedEventsCS);
+	return UserId;
+}
+
+void FRH_AnalyticsProviderEventCache::SetCorrelationId(TOptional<FString> InCorrelationId)
+{
+	FScopeLock ScopedLock(&CachedEventsCS);
+	CorrelationId = MoveTemp(InCorrelationId);
+}
+
+TOptional<FString> FRH_AnalyticsProviderEventCache::GetCorrelationId() const
+{
+	FScopeLock ScopedLock(&CachedEventsCS);
+	return CorrelationId;
+}
+
+/*
+We start with 
+{
+	"event_list":
+	[
+	]
+}
+
+We End with 
+{
+	"event_list":
+	[
+		{
+			"event_name":"<NAME>"
+			,"event_uuid":"<guid>"
+			,"event_timestamp":"<timestamp>"
+			,"correlation_id":"<str>"
+			,"user_id":"<str>"
+			"eventParams":
+			{
+				 <DefaultAttrs>
+				,<Attrs>
+			}
+			"custom_data":
+			{
+				<CustomDefaultAttrs>
+				,<CustomAttrs>
+			}
+	]
+}
+*/
 void FRH_AnalyticsProviderEventCache::AddToCache(FString EventName, const TArray<FAnalyticsEventAttribute>& Attributes)
 {
 	FScopeLock ScopedLock(&CachedEventsCS);
 
+	int32 CustomAttributeCount = RHEventCacheStatic::GetNumCustomAttributes(Attributes);
+
+	auto ComputeEventSize = [&]() -> int32
+		{
+			int32 FinalSize = 0;
+
+			{
+				const int32 BaseSize = 0
+					// {"event_name":"EVENT_NAME"
+					+ 14 + (1 + EventName.Len() + 1)
+					// ,"event_uuid":"00000000-0000-0000-0000-000000000000"
+					+ 14 + (1 + 36 + 1)
+					// ,"event_timestamp":"YYYY-MM-DDTHH:MM:SS.MMMZ"
+					+ 19 + (1 + 24 + 1)
+					// "},"    NOTE - closes out entire event after all other elements
+					+ 2
+					;
+				FinalSize += BaseSize;
+			}
+
+			if (CorrelationId.IsSet())
+			{
+				const int32 CorrelationSize = 0
+					// ,"correlation_id":"<string>"
+					+ 18 + (1 + CorrelationId.GetValue().Len() + 1)
+					;
+
+				FinalSize += CorrelationSize;
+			}
+
+			if (UserId.IsSet())
+			{
+				const int32 UserSize = 0
+					// ,"correlation_id":"<string>"
+					+ 18 + (1 + UserId.GetValue().Len() + 1)
```

Comment:
Is this length wrong for `,"user_id":"`?

Reply:
Bad copy/paste - the comment above is also incorrect.  Will fix